### PR TITLE
Simplex algorithm 

### PR DESCRIPTION
**Pull request by imengus**
### Describe your change:
Fixes: #8820
Implements the simplex algorithm for solving linear programs.
Adds `linear_programming` folder.

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [ ] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: linear_programming/simplex.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+        ... [1,0,-0.125,0.375,1]
+        ... ])
+        >>> t = Tableau(tableau, 2)
+        >>> t.interpret_tableau(tableau, ["x1", "x2", "s1", "s2", "RHS"])
+        {'P': 5.0, 'x1': 1.0, 'x2': 1.0}
+        """
+        output_dict = {}
+
+        # P = RHS of final tableau
+        output_dict["P"] = abs(tableau[0, -1])
```
Comment (CaedenPH) [inline linear_programming/simplex.py:301]:
```suggestion
        # P = RHS of final tableau
        output_dict = {"P":  abs(tableau[0, -1])}
```

File: linear_programming/simplex.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+        ['x1', 'x2', 's1', 's2', 's3', 'a1', 'RHS']
+        """
+        # decision | slack | artificial
+        string_starts = ["x", "s", "a"]
+        titles = []
+        for i in range(3):
+            for j in range(args[i]):
+                titles.append(string_starts[i] + str(j + 1))
+        titles.append("RHS")
+        return titles
```
Comment (CaedenPH) [inline linear_programming/simplex.py:88]: Much more consise
```suggestion
        return [string_starts[i] + str(j + 1) for j in range(args[i]) for i in range(3), "RHS"]
```

**Review by CaedenPH — CHANGES_REQUESTED** [review]
1) Reformat all your doctests so that the `np.ndarray` utilises the full line length, as shown in one suggestion
2) In what scenarioes would you enter a different `tableau` into `pivot`, `find_pivot`, and other functions like this? In the doctests you just us `t.tableau` but if you refactor this to not accept a tableau input you can create more concise doctests such as `Tableau(np.array([[-2,-1,0,0,0], [3,1,1,0,6], [1,2,0,1,7.]]), 2).find_pivot()` without having to assign the variable `t`

File: linear_programming/simplex.py (hunk 2 above)
Comment (imengus) [reply linear_programming/simplex.py:84]: This list comprehension doesn't work. For readability's sake, I'll stick with the for loops

**Comment by imengus** [comment]
@cclauss here is the simplex algorithm I mentioned in the issue. Please could you review?

File: linear_programming/simplex.py [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+            # If there are no more negative values in objective row
+            if self.stop_iter:
+                # Delete artificial variable columns and rows. Update attributes
+                self.tableau = self.change_stage(self.tableau)
+
+                # Pivot again
+                continue
+
+            self.tableau = self.pivot(self.tableau, row_idx, col_idx)
+            iter_num += 1
```
Comment (tianyizheng02) [inline linear_programming/simplex.py:274]:
```suggestion
        for _ in range(100):
            # Completion of each stage removes an objective. If both stages
            # are complete, then no objectives are left
            if not self.objectives:
                self.col_titles = self.generate_col_titles(
                    self.n_vars, self.n_slack, self.n_art_vars
                )

                # Find the values of each variable at optimal solution
                return self.interpret_tableau(self.tableau, self.col_titles)

            row_idx, col_idx = self.find_pivot(self.tableau)

            # If there are no more negative values in objective row
            if self.stop_iter:
                # Delete artificial variable columns and rows. Update attributes
                self.tableau = self.change_stage(self.tableau)
            else:
                self.tableau = self.pivot(self.tableau, row_idx, col_idx)
```
For-loops are more efficient than while-loops in Python, and they're easier to read in this case, IMO. Also,  the value of `iter_num` is never actually used

**Review by tianyizheng02 — CHANGES_REQUESTED** [review]

File: linear_programming/simplex.py (hunk 3 above)
Comment (imengus) [reply linear_programming/simplex.py:273]: Yes, agreed and thank you

**Review by cclauss — APPROVED** [review]

**Pull request merged**